cpp/src/arrow/python/common.h

```diff
@@ -180,6 +180,9 @@ class ARROW_PYTHON_EXPORT OwnedRefNoGIL : public OwnedRef {
   explicit OwnedRefNoGIL(PyObject* obj) : OwnedRef(obj) {}
 
   ~OwnedRefNoGIL() {
+    if (obj() == NULLPTR) {
```

Comment:
Why are we creating a ref to nullptr in the first place?

Reply:
This is from the call to `detach`.  Concretely we had to make this change to avoid trying to acquire the GIL during/after finalization (by detecting that we are finalizing/finalized and detaching in `PythonUdf`).

However, since `OwnedRefNoGIL` is a moveable type, we are probably encountering this situation already (we detach on move) but getting away with it because it calls `Py_XDECREF` which tolerates null.  So this will hopefully prevent us from needlessly grabbing the GIL in a few other situations as well.

Reply:
Sure, I will add this.

Reply:
This particular change is not specific to finalization.  For example:

```
void Foo(PyObject* obj) {
    OwnedRefNoGIL obj_ref(obj);
    Bar(std::move(obj_ref));
    // At the close of this method `~OwnedRefNoGIL` will be called on `obj_ref` which is no longer valid.  We shouldn't
    // bother grabbing the GIL for no reason.
}
```

The part that is related to finalization is the `~PythonUdf` part (and we already have a comment there)`.

Reply:
Well, it's a correctness issue for finalization, while it's only an optimization otherwise, so I think the comment is important.